keps/sig-apps/5440-mutable-job-pod-resource-updates/README.md

```diff
@@ -0,0 +1,537 @@
+- Support in-place pod resource updates (this is covered by separate KEPs).
+- Allow mutating of Pod Resources.
+
+## Proposal
+
+The proposal is to relax update validation for container resource specifications
+(CPU, memory, GPU, and extended resource requests and limits) in the pod template of suspended jobs.
+
+This change has minimal impact on the job-controller, as the job controller will
+use the updated resource specifications when creating new pods for the job.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to build a controller that implements job queueing with dynamic resource optimization.
+Users create v1.Job objects, and to control when the job can run, I have a webhook that
+forces the jobs to be created in a suspended state. The controller analyzes current
+cluster capacity and adjusts job resource requirements to optimize cluster utilization
+before unsuspending them.
+
+At job creation time, users may specify conservative resource estimates or may not know
+the optimal resource allocation for current cluster conditions. The queue controller can
+analyze available capacity, other queued jobs, and cluster utilization patterns to
+determine optimal CPU, memory, and GPU allocations. For example, it might adjust the number of GPUs based on current
+availability. By updating the job's resource requirements before unsuspending it, the
+controller ensures efficient resource utilization and better cluster throughput.
+
+### Risks and Mitigations
+
+- New API calls from queue controllers to update resource specifications. The mitigation
+  is for such controllers to make a single API call for both updating resources and
+  unsuspending the job.
+  
+- Potential for resource specification changes to make a job unschedulable if the
+  updated requirements exceed available cluster capacity. Queue controllers should
+  validate resource availability before making changes.
+
+- A race condition could theoretically happen if a job is unsuspended and then quickly
+  suspended again before resource updates, though this is not a typical use case pattern.
+
+## Design Details
+
+The pod template validation logic in the API server needs to be updated to relax the validation
+of the Job's Template field. Currently the template is immutable, but we need to make
+container resource specifications (CPU, memory, GPU, and extended resources requests and limits) mutable for suspended jobs.
+
+The condition we will check to verify that the job is suspended is `Job.Spec.Suspend=true`.
+
+We will allow updates to the following fields in container specifications within the pod template:
+- `resources.requests.cpu`
+- `resources.requests.memory`
+- `resources.requests.*` (for extended resources like `nvidia.com/gpu`, `amd.com/gpu`, `tpu-v4` etc.)
+- `resources.limits.cpu`
+- `resources.limits.memory`
+- `resources.limits.*` (for extended resources like `nvidia.com/gpu`, `amd.com/gpu`, `tpu-v4` etc.)
```

Comment:
I don't see `resources.claims` listed here. Does that mean DRA is out of scope? Either way, could we clarify that in the goals/non-goals above?